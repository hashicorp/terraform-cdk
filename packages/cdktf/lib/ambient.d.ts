// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
declare module 'archiver';