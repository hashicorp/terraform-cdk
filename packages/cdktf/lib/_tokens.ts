// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, DefaultTokenResolver, StringConcat } from "./tokens";
import { Construct } from "constructs";

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

// eslint-disable-next-line jsdoc/require-jsdoc
export function resolve<T>(scope: Construct, obj: T, preparing = false): T {
  return Tokenization.resolve(obj, {
    scope,
    preparing,
    resolver: TOKEN_RESOLVER,
  });
}
