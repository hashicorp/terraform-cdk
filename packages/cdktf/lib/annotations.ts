// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
// Originally from aws-cdk v2 because with constructs v10 addMessage was removed
// in favor of Annotations in the AWS CDK
// https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/annotations.ts
import { IConstruct } from "constructs";

export enum AnnotationMetadataEntryType {
  INFO = "@cdktf/info",
  WARN = "@cdktf/warn",
  ERROR = "@cdktf/error",
}

export const DISABLE_STACK_TRACE_IN_METADATA =
  "@cdktf/disable_stack_trace_in_metadata";

/**
 * Includes API for attaching annotations such as warning messages to constructs.
 */
export class Annotations {
  /**
   * Returns the annotations API for a construct scope.
   * @param scope The scope
   */
  public static of(scope: IConstruct) {
    return new Annotations(scope);
  }

  private readonly stackTraces: boolean;

  private constructor(private readonly scope: IConstruct) {
    const disableTrace =
      scope.node.tryGetContext(DISABLE_STACK_TRACE_IN_METADATA) || false;

    this.stackTraces = !disableTrace;
  }

  /**
   * Adds a warning metadata entry to this construct.
   *
   * The CLI will display the warning when an app is synthesized.
   * In a future release the CLI might introduce a --strict flag which
   * will then fail the synthesis if it encounters a warning.
   *
   * @param message The warning message.
   */
  public addWarning(message: string) {
    this.addMessage(AnnotationMetadataEntryType.WARN, message);
  }

  /**
   * Adds an info metadata entry to this construct.
   *
   * The CLI will display the info message when apps are synthesized.
   *
   * @param message The info message.
   */
  public addInfo(message: string): void {
    this.addMessage(AnnotationMetadataEntryType.INFO, message);
  }

  /**
   * Adds an { "error": <message> } metadata entry to this construct.
   * The toolkit will fail synthesis when errors are reported.
   * @param message The error message.
   */
  public addError(message: string) {
    this.addMessage(AnnotationMetadataEntryType.ERROR, message);
  }

  /**
   * Adds a message metadata entry to the construct node, to be displayed by the CDK CLI.
   * @param level The message level
   * @param message The message itself
   */
  private addMessage(level: AnnotationMetadataEntryType, message: string) {
    this.scope.node.addMetadata(level, message, {
      stackTrace: this.stackTraces,
    });
  }
}
