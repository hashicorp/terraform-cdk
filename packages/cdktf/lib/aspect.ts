// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
//  Originally from aws-cdk v2 because with constructs v10 Aspects where moved to the AWS CDK
//  https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/aspect.ts
import { IConstruct } from "constructs";

const ASPECTS_SYMBOL = Symbol("cdktf-aspects");

/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * All aspects can visit an IConstruct
   */
  visit(node: IConstruct): void;
}

/**
 * Aspects can be applied to CDK tree scopes and can operate on the tree before
 * synthesis.
 */
export class Aspects {
  /**
   * Returns the `Aspects` object associated with a construct scope.
   * @param scope The scope for which these aspects will apply.
   */
  public static of(scope: IConstruct): Aspects {
    let aspects = (scope as any)[ASPECTS_SYMBOL];
    if (!aspects) {
      aspects = new Aspects();

      Object.defineProperty(scope, ASPECTS_SYMBOL, {
        value: aspects,
        configurable: false,
        enumerable: false,
      });
    }
    return aspects;
  }

  private readonly _aspects: IAspect[];

  private constructor() {
    this._aspects = [];
  }

  /**
   * Adds an aspect to apply this scope before synthesis.
   * @param aspect The aspect to add.
   */
  public add(aspect: IAspect) {
    this._aspects.push(aspect);
  }

  /**
   * The list of aspects which were directly applied on this scope.
   */
  public get all(): IAspect[] {
    return [...this._aspects];
  }
}
