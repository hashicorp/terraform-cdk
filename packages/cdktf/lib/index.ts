// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { validateEnvironment } from "./validateEnvironment";
validateEnvironment();

export * from "./tokens";
export * from "./terraform-stack";
export * from "./terraform-element";
export * from "./terraform-module";
export * from "./terraform-provider";
export * from "./terraform-resource";
export * from "./terraform-data-source";
export * from "./terraform-output";
export * from "./complex-computed-list";
export * from "./resource";
export * from "./testing";
export * as testingMatchers from "./testing/matchers";
export * from "./app";
export * from "./terraform-backend";
export * from "./backends";
export * from "./terraform-remote-state";
export * from "./terraform-local";
export * from "./terraform-variable";
export * from "./terraform-hcl-module";
export * from "./runtime";
export * from "./terraform-dependable";
export * from "./manifest";
export * from "./terraform-asset";
export * from "./synthesize";
export * from "./annotations";
export * from "./aspect";
export * from "./terraform-functions";
export * from "./terraform-operators";
export * from "./tfExpression";
export * from "./terraform-addressable";
export * from "./terraform-iterator";
export * from "./terraform-provisioner";
export * from "./terraform-conditions";
export * from "./terraform-count";

// required for JSII because Fn extends from it
export * from "./functions/terraform-functions.generated";
