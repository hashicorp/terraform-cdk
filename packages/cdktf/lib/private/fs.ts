// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import * as crypto from "crypto";

const HASH_LEN = 32;

// Full implementation at https://github.com/jprichardson/node-fs-extra/blob/master/lib/copy/copy-sync.js
/**
 * Copy a file or directory. The directory can have contents and subfolders.
 * @param {string} src
 * @param {string} dest
 */
export function copySync(src: string, dest: string) {
  /**
   * Copies file if present otherwise walks subfolder
   * @param {string} p
   */
  function copyItem(p: string) {
    const sourcePath = path.resolve(src, p);
    const stat = fs.statSync(sourcePath);
    if (stat.isFile()) {
      fs.copyFileSync(sourcePath, path.resolve(dest, p));
    }
    if (stat.isDirectory()) {
      walkSubfolder(p);
    }
  }
  /**
   * Copies contents of subfolder
   * @param {string} p
   */
  function walkSubfolder(p: string) {
    const sourceDir = path.resolve(src, p);
    fs.mkdirSync(path.resolve(dest, p), { recursive: true });
    fs.readdirSync(sourceDir).forEach((item) => copyItem(path.join(p, item)));
  }

  walkSubfolder(".");
}

/**
 * Zips contents at src and places zip archive at dest
 * @param {string} src
 * @param {string} dest
 */
export function archiveSync(src: string, dest: string) {
  const projectRoot = path.resolve(__dirname, "..", "..");
  const zipSyncPath = path.resolve(projectRoot, "bin", "zipSync.js");
  execSync(`node ${zipSyncPath} ${src} ${dest}`);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function hashPath(src: string) {
  const hash = crypto.createHash("md5");

  // eslint-disable-next-line jsdoc/require-jsdoc
  function hashRecursion(p: string) {
    const stat = fs.statSync(p);
    if (stat.isFile()) {
      hash.update(fs.readFileSync(p));
    } else if (stat.isDirectory()) {
      fs.readdirSync(p).forEach((filename) =>
        hashRecursion(path.resolve(p, filename))
      );
    }
  }

  hashRecursion(src);
  return hash.digest("hex").slice(0, HASH_LEN).toUpperCase();
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function findFileAboveCwd(
  file: string,
  rootPath = process.cwd()
): string | null {
  const fullPath = path.resolve(rootPath, file);
  if (fs.existsSync(fullPath)) {
    return fullPath;
  }

  const parentDir = path.resolve(rootPath, "..");
  if (fs.existsSync(parentDir) && parentDir !== rootPath) {
    return findFileAboveCwd(file, parentDir);
  }

  return null;
}
