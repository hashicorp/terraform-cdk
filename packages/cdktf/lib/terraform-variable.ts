// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
import { Construct } from "constructs";
import { TerraformElement } from "./terraform-element";
import { keysToSnakeCase, deepMerge } from "./util";
import { Token } from "./tokens";
import { Expression, ref } from "./tfExpression";
import { IResolvable } from "./tokens/resolvable";
import { ITerraformAddressable } from "./terraform-addressable";

// eslint-disable-next-line jsdoc/require-jsdoc
export abstract class VariableType {
  public static readonly STRING = "string";
  public static readonly NUMBER = "number";
  public static readonly BOOL = "bool";
  public static readonly ANY = "any";

  public static readonly LIST = "list";
  public static readonly MAP = "map";
  public static readonly SET = "set";

  public static readonly LIST_STRING = "list(string)";
  public static readonly LIST_NUMBER = "list(number)";
  public static readonly LIST_BOOL = "list(bool)";

  public static readonly MAP_STRING = "map(string)";
  public static readonly MAP_NUMBER = "map(number)";
  public static readonly MAP_BOOL = "map(bool)";

  public static readonly SET_STRING = "set(string)";
  public static readonly SET_NUMBER = "set(number)";
  public static readonly SET_BOOL = "set(bool)";

  public static list(type: string) {
    return `list(${type})`;
  }

  public static map(type: string) {
    return `map(${type})`;
  }

  public static set(type: string) {
    return `set(${type})`;
  }

  public static tuple(...elements: string[]) {
    return `tuple(${elements.join(", ")})`;
  }

  public static object(attributes: { [key: string]: string }) {
    return `object({${Object.keys(attributes)
      .map((k) => k + "=" + attributes[k])
      .join(", ")}})`;
  }
}

export interface TerraformVariableValidationConfig {
  readonly condition: Expression;
  readonly errorMessage: string;
}

export interface TerraformVariableConfig {
  readonly default?: any;
  readonly description?: string;

  /**
   * The type argument in a variable block allows you to restrict the type of value that will be accepted as the value for a variable. If no type constraint is set then a value of any type is accepted.
   *
   * While type constraints are optional, we recommend specifying them; they serve as easy reminders for users of the module, and allow Terraform to return a helpful error message if the wrong type is used.
   *
   * Type constraints are created from a mixture of type keywords and type constructors. The supported type keywords are:
   *
   * - string
   * - number
   * - bool
   *
   * The type constructors allow you to specify complex types such as collections:
   *
   * - list(\<TYPE\>)
   * - set(\<TYPE\>)
   * - map(\<TYPE\>)
   * - object({\<ATTR NAME\> = \<TYPE\>, ... })
   * - tuple([\<TYPE\>, ...])
   *
   * The keyword any may be used to indicate that any type is acceptable. For more information on the meaning and behavior of these different types, as well as detailed information about automatic conversion of complex types, see {@link https://www.terraform.io/docs/configuration/types.html|Type Constraints}.
   *
   * If both the type and default arguments are specified, the given default value must be convertible to the specified type.
   */
  readonly type?: string;

  readonly sensitive?: boolean;

  /*
   * The nullable argument in a variable block controls whether the module caller may assign the value null to the variable.
   */
  readonly nullable?: boolean;

  /**
   * Specify arbitrary custom validation rules for a particular variable using a validation block nested within the corresponding variable block
   */
  readonly validation?: TerraformVariableValidationConfig[];
}

// eslint-disable-next-line jsdoc/require-jsdoc
export class TerraformVariable
  extends TerraformElement
  implements ITerraformAddressable
{
  public readonly default?: any;
  public readonly description?: string;
  public readonly type?: string;
  public readonly sensitive?: boolean;
  public readonly nullable?: boolean;
  private _validation?: TerraformVariableValidationConfig[];

  constructor(scope: Construct, id: string, config: TerraformVariableConfig) {
    super(scope, id, "var");

    this.default = config.default;
    this.description = config.description;
    this.type = config.type;
    this.sensitive = config.sensitive;
    this.nullable = config.nullable;
    this._validation = config.validation;
  }

  public get stringValue(): string {
    return Token.asString(this.interpolation());
  }

  public get numberValue(): number {
    return Token.asNumber(this.interpolation());
  }

  public get listValue(): string[] {
    return Token.asList(this.interpolation());
  }

  public get booleanValue(): IResolvable {
    return this.interpolation();
  }

  public get value(): any {
    return Token.asAny(this.interpolation());
  }

  public get validation(): TerraformVariableValidationConfig[] | undefined {
    return this._validation;
  }

  public addValidation(validation: TerraformVariableValidationConfig) {
    if (!this._validation) {
      this._validation = [];
    }

    this._validation.push(validation);
  }

  private interpolation(): IResolvable {
    return ref(`var.${this.friendlyUniqueId}`, this.cdktfStack);
  }

  public synthesizeAttributes(): { [key: string]: any } {
    return {
      default: this.default,
      description: this.description,
      type: this.type,
      sensitive: this.sensitive,
      nullable: this.nullable,
      validation: this.validation?.map((validation) => ({
        error_message: validation.errorMessage,
        condition: validation.condition,
      })),
    };
  }

  public toTerraform(): any {
    return {
      variable: {
        [this.friendlyUniqueId]: deepMerge(
          keysToSnakeCase(this.synthesizeAttributes()),
          this.rawOverrides
        ),
      },
    };
  }
}
