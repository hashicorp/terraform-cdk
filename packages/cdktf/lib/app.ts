// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Construct, IConstruct } from "constructs";
import * as fs from "fs";
import { version } from "../package.json";
import { DISABLE_STACK_TRACE_IN_METADATA } from "./annotations";
import { Manifest } from "./manifest";
import { ISynthesisSession } from "./synthesize";
import { TerraformStack } from "./terraform-stack";

const APP_SYMBOL = Symbol.for("cdktf/App");
export const CONTEXT_ENV = "CDKTF_CONTEXT_JSON";
export interface AppConfig {
  /**
   * The directory to output Terraform resources.
   *
   * @default - CDKTF_OUTDIR if defined, otherwise "cdktf.out"
   */
  readonly outdir?: string;
  readonly stackTraces?: boolean;

  /**
   * Additional context values for the application.
   *
   * Context set by the CLI or the `context` key in `cdktf.json` has precedence.
   *
   * Context can be read from any construct using `node.getContext(key)`.
   *
   * @default - no additional context
   */
  readonly context?: { [key: string]: any };

  /**
   * Whether to skip all validations during synthesis of the app
   *
   * @default - false
   */
  readonly skipValidation?: boolean;

  /**
   * Whether to skip backend validation during synthesis of the app
   *
   * @default - false
   */
  readonly skipBackendValidation?: boolean;
}

/**
 * Represents a cdktf application.
 */
export class App extends Construct {
  /**
   * The output directory into which resources will be synthesized.
   */
  public readonly outdir: string;

  /**
   * The stack which will be synthesized. If not set, all stacks will be synthesized.
   */
  public readonly targetStackId: string | undefined;

  public readonly manifest: Manifest;

  /**
   * Whether to skip all validations during synthesis of the app
   */
  public readonly skipValidation: boolean;

  /**
   * Whether to skip backend validation during synthesis of the app
   */
  public readonly skipBackendValidation: boolean;

  /**
   * Defines an app
   * @param config configuration
   */
  constructor(config: AppConfig = {}) {
    super(undefined as any, "");
    Object.defineProperty(this, APP_SYMBOL, { value: true });

    this.outdir = process.env.CDKTF_OUTDIR ?? config.outdir ?? "cdktf.out";
    this.targetStackId = process.env.CDKTF_TARGET_STACK_ID;
    this.skipValidation = config.skipValidation || false;
    this.skipBackendValidation = config.skipBackendValidation || false;

    this.loadContext(config.context);

    const node = this.node;
    if (config.stackTraces === false) {
      node.setContext(DISABLE_STACK_TRACE_IN_METADATA, true);
    }

    node.setContext("cdktfVersion", version);

    if (!fs.existsSync(this.outdir)) {
      fs.mkdirSync(this.outdir);
    }
    this.manifest = new Manifest(version, this.outdir);
  }

  public static isApp(x: any): x is App {
    return x !== null && typeof x === "object" && APP_SYMBOL in x;
  }

  public static of(construct: IConstruct): App {
    return _lookup(construct);

    // eslint-disable-next-line jsdoc/require-jsdoc
    function _lookup(c: IConstruct): App {
      if (App.isApp(c)) {
        return c;
      }

      const node = c.node;

      if (!node.scope) {
        throw new Error(
          `No app could be identified for the construct at path '${construct.node.path}'`
        );
      }

      return _lookup(node.scope);
    }
  }

  /**
   * Synthesizes all resources to the output directory
   */
  public synth(): void {
    const session: ISynthesisSession = {
      outdir: this.outdir,
      skipValidation: this.skipValidation,
      manifest: this.manifest,
    };

    const stacks = this.node
      .findAll()
      .filter<TerraformStack>(TerraformStack.isStack);

    stacks.forEach((stack) => stack.prepareStack());
    stacks.forEach((stack) => stack.synthesizer.synthesize(session));

    if (!this.skipValidation) {
      const validations = this.node.validate();
      if (validations.length) {
        const errorList = validations.join("\n  ");
        throw new Error(
          `App level validation failed with the following errors:\n  ${errorList}`
        );
      }
    }

    this.manifest.writeToFile();
  }

  /**
   * Creates a reference from one stack to another, invoked on prepareStack since it creates extra resources
   */
  public crossStackReference(
    fromStack: TerraformStack,
    toStack: TerraformStack,
    identifier: string
  ): string {
    toStack.addDependency(fromStack);
    const outputId =
      fromStack.registerOutgoingCrossStackReference(
        identifier
      ).friendlyUniqueId;

    const remoteState = toStack.registerIncomingCrossStackReference(fromStack);

    return remoteState.getString(outputId);
  }

  private loadContext(defaults: { [key: string]: string } = {}) {
    const node = this.node;

    // prime with defaults passed through constructor
    for (const [k, v] of Object.entries(defaults)) {
      node.setContext(k, v);
    }

    // read from environment
    const contextJson = process.env[CONTEXT_ENV];
    const contextFromEnvironment = contextJson ? JSON.parse(contextJson) : {};

    for (const [k, v] of Object.entries(contextFromEnvironment)) {
      node.setContext(k, v);
    }
  }
}
