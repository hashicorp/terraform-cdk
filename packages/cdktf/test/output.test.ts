// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import {
  Testing,
  TerraformStack,
  TerraformOutput,
  TerraformElement,
  App,
  Fn,
} from "../lib";
import fs = require("fs");
import path = require("path");
import os = require("os");
import { TestResource, TestProvider } from "./helper";
import { TerraformVariable } from "../lib/terraform-variable";
import { IConstruct } from "constructs";

test("number output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: 1,
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("string output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: "1",
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("boolean output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: true,
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("map output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: { foo: "bar" },
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("list output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: ["foo", "bar"],
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("description output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: 1,
    description: "test-description",
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("sensitive output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: 1,
    sensitive: true,
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("dependent output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TestProvider(stack, "provider", {});
  const resource = new TestResource(stack, "weird-long-running-resource", {
    name: "foo",
  });

  new TerraformOutput(stack, "test-output", {
    value: 1,
    dependsOn: [resource],
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("variable output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-variable", {});

  new TerraformOutput(stack, "test-output", {
    value: variable.value,
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("static output id (without feature flags enabled)", () => {
  // we do this manually instead of using Testing.app() to skip enabling feature flags
  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), "cdktf.outdir."));
  const app = Testing.stubVersion(new App({ outdir, stackTraces: false }));

  const stack = new TerraformStack(app, "test");
  class CustomConstruct extends TerraformElement {
    constructor(scope: IConstruct, id: string) {
      super(scope, id);
      new TerraformOutput(stack, "my_static_output_name", {
        value: "1",
        staticId: true,
      });
    }
  }
  new CustomConstruct(
    stack,
    "custom-construct-id-which-does-not-appear-in-output-name"
  );

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("full resource output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TestProvider(stack, "provider", {});
  const resource = new TestResource(stack, "foo", {
    name: "foo",
  });

  new TerraformOutput(stack, "test-output", {
    value: resource,
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("expression output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TestProvider(stack, "provider", {});
  const resource = new TestResource(stack, "foo", {
    name: "foo",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.upper(resource.name),
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("resource[] output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TestProvider(stack, "provider", {});
  const resource1 = new TestResource(stack, "foo1", {
    name: "foo1",
  });
  const resource2 = new TestResource(stack, "foo2", {
    name: "foo2",
  });

  new TerraformOutput(stack, "test-output", {
    value: [resource1, resource2],
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("resource map output", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TestProvider(stack, "provider", {});
  const resource = new TestResource(stack, "foo", {
    name: "foo",
  });
  new TerraformOutput(stack, "test-output", {
    value: {
      myResource: resource,
      something: "else",
    },
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});
