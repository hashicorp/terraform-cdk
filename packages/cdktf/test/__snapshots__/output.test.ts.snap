// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boolean output 1`] = `
"{
  "output": {
    "test-output": {
      "value": true
    }
  }
}"
`;

exports[`dependent output 1`] = `
"{
  "output": {
    "test-output": {
      "depends_on": [
        "\${test_resource.weird-long-running-resource}"
      ],
      "value": 1
    }
  },
  "provider": {
    "test": [
      {
      }
    ]
  },
  "resource": {
    "test_resource": {
      "weird-long-running-resource": {
        "name": "foo"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "test": {
        "version": "~> 2.0"
      }
    }
  }
}"
`;

exports[`description output 1`] = `
"{
  "output": {
    "test-output": {
      "description": "test-description",
      "value": 1
    }
  }
}"
`;

exports[`expression output 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${upper(\\"foo\\")}"
    }
  },
  "provider": {
    "test": [
      {
      }
    ]
  },
  "resource": {
    "test_resource": {
      "foo": {
        "name": "foo"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "test": {
        "version": "~> 2.0"
      }
    }
  }
}"
`;

exports[`full resource output 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${test_resource.foo}"
    }
  },
  "provider": {
    "test": [
      {
      }
    ]
  },
  "resource": {
    "test_resource": {
      "foo": {
        "name": "foo"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "test": {
        "version": "~> 2.0"
      }
    }
  }
}"
`;

exports[`list output 1`] = `
"{
  "output": {
    "test-output": {
      "value": [
        "foo",
        "bar"
      ]
    }
  }
}"
`;

exports[`map output 1`] = `
"{
  "output": {
    "test-output": {
      "value": {
        "foo": "bar"
      }
    }
  }
}"
`;

exports[`number output 1`] = `
"{
  "output": {
    "test-output": {
      "value": 1
    }
  }
}"
`;

exports[`resource map output 1`] = `
"{
  "output": {
    "test-output": {
      "value": {
        "myResource": "\${test_resource.foo}",
        "something": "else"
      }
    }
  },
  "provider": {
    "test": [
      {
      }
    ]
  },
  "resource": {
    "test_resource": {
      "foo": {
        "name": "foo"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "test": {
        "version": "~> 2.0"
      }
    }
  }
}"
`;

exports[`resource[] output 1`] = `
"{
  "output": {
    "test-output": {
      "value": [
        "\${test_resource.foo1}",
        "\${test_resource.foo2}"
      ]
    }
  },
  "provider": {
    "test": [
      {
      }
    ]
  },
  "resource": {
    "test_resource": {
      "foo1": {
        "name": "foo1"
      },
      "foo2": {
        "name": "foo2"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "test": {
        "version": "~> 2.0"
      }
    }
  }
}"
`;

exports[`sensitive output 1`] = `
"{
  "output": {
    "test-output": {
      "sensitive": true,
      "value": 1
    }
  }
}"
`;

exports[`static output id (without feature flags enabled) 1`] = `
"{
  "output": {
    "my_static_output_name": {
      "value": "1"
    }
  }
}"
`;

exports[`string output 1`] = `
"{
  "output": {
    "test-output": {
      "value": "1"
    }
  }
}"
`;

exports[`variable output 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${var.test-variable}"
    }
  },
  "variable": {
    "test-variable": {
    }
  }
}"
`;
